src/main/java/com/icthh/xm/ms/entity/service/impl/FunctionServiceImpl.java

```diff
@@ -79,40 +83,73 @@ public FunctionContext execute(final String functionKey,
 
         Map<String, Object> vInput = CustomCollectionUtils.emptyIfNull(functionInput);
 
+        dynamicPermissionCheckService.checkContextPermission(DynamicPermissionCheckService.FeatureContext.FUNCTION,
+            "XMENTITY.FUNCTION.EXECUTE", functionKey);
+
         // get type key
-        XmEntityIdKeyTypeKey projection = xmEntityService.getXmEntityIdKeyTypeKey(idOrKey);
-        String xmEntityTypeKey = projection.getTypeKey();
+        XmEntityStateProjection projection = xmEntityService.findStateProjectionById(idOrKey).orElseThrow(
+            () -> new EntityNotFoundException("XmEntity with idOrKey [" + idOrKey + "] not found")
+        );
 
         // validate that current XmEntity has function
-        FunctionSpec functionSpec = xmEntitySpecService.findFunction(xmEntityTypeKey, functionKey).orElseThrow(
-            () -> new IllegalArgumentException("Function not found for entity type key " + xmEntityTypeKey
-                + " and function key: " + functionKey)
-        );
+        FunctionSpec functionSpec = findFunctionSpec(functionKey, projection);
+
+        isCallAllowedByState(functionSpec, projection).orElseThrow(
+            () -> new IllegalStateException("Function call forbidden for current state"));
```

Comment:
we have allowed state key filed in specification.
this is related to that field?